utils/terraform_client.py

```diff
@@ -0,0 +1,143 @@
+        )
+
+
+class OR(OpenshiftResource):
+    def __init__(self, body, integration, integration_version):
+        super(OR, self).__init__(
+            body, integration, integration_version
+        )
+
+
+class TerraformClient(object):
+    def __init__(self, integration, integration_version,
+                 integration_prefix, working_dirs):
+        self.integration = integration
+        self.integration_version = integration_version
+        self.integration_prefix = integration_prefix
+        self.working_dirs = working_dirs
+        tfs = {}
+        for name, wd in working_dirs.items():
+            tf = Terraform(working_dir=wd)
+            self.return_code, self.stdout, self.stderr = tf.init()
+            self.check_output()
+            tfs[name] = tf
+        self.tfs = tfs
+
+    def plan(self):
+        for name, tf in self.tfs.items():
+            self.return_code, self.stdout, self.stderr = \
+                tf.plan(detailed_exitcode=False)
+            self.check_output()
+            self.log_plan_diff(name)
+
+    def log_plan_diff(self, name):
+        for line in self.stdout.split('\n'):
+            line = line.strip()
+            if line.startswith('+ aws'):
+                line_split = line.replace('+ ', '').split('.')
+                logging.info(['create', name, line_split[0], line_split[1]])
+            if line.startswith('- aws'):
+                line_split = line.replace('- ', '').split('.')
+                logging.info(['destroy', name, line_split[0], line_split[1]])
+            if line.startswith('~ aws'):
+                line_split = line.replace('~ ', '').split('.')
+                logging.info(['update', name, line_split[0], line_split[1]])
+            if line.startswith('-/+ aws'):
+                line_split = line.replace('-/+ ', '').split('.')
+                logging.info(['REPLACE', name, line_split[0],
+                              line_split[1].split(' ', 1)[0]])
+
+    def apply(self):
+        for name, tf in self.tfs.items():
+            self.return_code, self.stdout, self.stderr = \
+                tf.apply(auto_approve=True)
+            self.check_output()
+
+    def populate_desired_state(self, ri):
+        for name, tf in self.tfs.items():
+            self.output = tf.output()
+            self.format_output()
+
+            for name, data in self.output.items():
+                oc_resource = self.contruct_oc_resource(name, data)
+                ri.add_desired(
+                    data['{}.cluster'.format(self.integration_prefix)],
+                    data['{}.namespace'.format(self.integration_prefix)],
+                    data['{}.resource'.format(self.integration_prefix)],
+                    name,
+                    oc_resource
+                )
+
+    def format_output(self):
+        # data is a dictionary of dictionaries
+        data = {}
+        for k, v in self.output.items():
+            if '[' not in k or ']' not in k:
```

Comment:
we are protecting ourselves against any output that may be added during debugging or manual actions. the integration will store all its output in the form of `<openshift_secret_to_be_created>[<key>] = <value>`. in case we perform manual actions, they are (potentially) kept in the state, and may be returned at a later time. this is how we prevent this from happening.

Reply:
can you please add this as a document

Reply:
sure